system/lib/pthread/pthread_kill.c

```diff
@@ -5,11 +5,28 @@
  * found in the LICENSE file.
  */
 
-#include <emscripten/threading.h>
+#include <threading_internal.h>
+#include <emscripten/proxying.h>
+#include <emscripten/console.h>
+#include <emscripten_internal.h>
 
 #include "pthread_impl.h"
 #include "lock.h"
 
+void do_raise(void* arg) {
+  int sig = (intptr_t)arg;
+  if (sig == SIGCANCEL) {
+    // For SIGCANCEL is simply enough to run the proxied function since we call
+    // pthread_testcancel at the end of _emscripten_check_mailbox.  We
+    // can't/don't call pthread_testcancel or pthread_exit here because the
+    // proxying system itself is on the stack and we want to exit in clean
+    // state. (e.g. without holding any locks).
+    _emscripten_runtime_keepalive_clear();
```

Comment:
Does "run the proxied function" refer to `_emscripten_runtime_keepalive_clear`? In `action_terminate()` in `raise.c` I see we also do `_Exit(128 + sig);`

Reply:
`SIGCANCEL` doesn't want to terminate the whole process, only the current thread.    Its a special internal signal that effects just the current thread.

Reply:
I updated this comment.